### PR TITLE
Add clinic code to End User view on Wagtail admin; allow for downloading of CSV

### PR DESCRIPTION
**Pull request by phala**

### CODE REVIEW

File: nurseconnect/admin.py [hunk 1]
```diff
@@ -47,3 +54,36 @@ class YourWordsModelAdminGroup(ModelAdminGroup):
     menu_order = 300
     items = (MoloYourWordsCompetitionModelAdmin,
              MoloYourWordsCompetitionEntryModelAdmin)
+
+
+def download_as_csv(NurseConnectEndUsersModelAdmin, request, queryset):
+    response = HttpResponse(content_type='text/csv')
+    response['Content-Disposition'] = 'attachment;filename=export.csv'
```
Comment (chrislombaard) [inline nurseconnect/admin.py]: Double quotes please.

File: nurseconnect/admin.py [hunk 2]
```diff
@@ -47,3 +54,36 @@ class YourWordsModelAdminGroup(ModelAdminGroup):
     menu_order = 300
     items = (MoloYourWordsCompetitionModelAdmin,
              MoloYourWordsCompetitionEntryModelAdmin)
+
+
+def download_as_csv(NurseConnectEndUsersModelAdmin, request, queryset):
+    response = HttpResponse(content_type='text/csv')
```
Comment (chrislombaard) [inline nurseconnect/admin.py]: Double quotes please.

File: nurseconnect/admin.py [hunk 3]
```diff
@@ -47,3 +54,36 @@ class YourWordsModelAdminGroup(ModelAdminGroup):
...
+    response['Content-Disposition'] = 'attachment;filename=export.csv'
+    writer = csv.writer(response)
+    field_names = NurseConnectEndUsersModelAdmin.list_display
+    writer.writerow(field_names)
+    for obj in queryset:
+        obj.username = obj.username.encode('utf-8')
```
Comment (chrislombaard) [inline nurseconnect/admin.py]: DQ

File: nurseconnect/admin.py [hunk 4]
```diff
@@ -47,3 +54,36 @@ class YourWordsModelAdminGroup(ModelAdminGroup):
...
+        obj.date_joined = obj.date_joined.strftime("%Y-%m-%d %H:%M")
+        writer.writerow(
+            [getattr(obj, field) for field in field_names]
+        )
+    return response
+download_as_csv.short_description = "Download selected as csv"
+
+
+class NurseConnectEndUsersModelAdmin(FrontendUsersModelAdmin):
+    menu_label = 'End Users'
```
Comment (chrislombaard) [inline nurseconnect/admin.py]: DQ

File: nurseconnect/admin.py [hunk 5]
```diff
@@ -47,3 +54,36 @@ class YourWordsModelAdminGroup(ModelAdminGroup):
...
+    menu_icon = 'user'
```
Comment (chrislombaard) [inline nurseconnect/admin.py]: DQ

File: nurseconnect/admin_import_export.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+from import_export import resources
+
+from django.contrib.auth.models import User
+from import_export.fields import Field
```
Comment (chrislombaard) [inline nurseconnect/admin_import_export.py:4]: Group import statements according parent app

File: nurseconnect/admin_import_export.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    Adapted from molo.profiles.
+    NurseConnect requires customization of this feature.
+    `clinic_code` field also has to be available for export.
+    """
+    # see dehydrate_ functions below
+    clinic_code = Field()
+
+    class Meta:
+        model = User
+
```
Comment (chrislombaard) [inline nurseconnect/admin_import_export.py:19]: Remove space

File: nurseconnect/admin_import_export.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        exclude = ('id', 'password', 'is_superuser', 'groups',
+                   'user_permissions', 'is_staff')
+
```
Comment (chrislombaard) [inline nurseconnect/admin_import_export.py:22]: Remove space

File: nurseconnect/admin_import_export.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        fields = ("username", "clinic_code")
+
```
Comment (chrislombaard) [inline nurseconnect/admin_import_export.py:24]: Remove space

File: nurseconnect/admin_import_export.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    NurseConnect requires customization of this feature.
+    `clinic_code` field also has to be available for export.
+    """
+    # see dehydrate_ functions below
+    clinic_code = Field()
+
+    class Meta:
+        model = User
+
+        exclude = ('id', 'password', 'is_superuser', 'groups',
```
Comment (chrislombaard) [inline nurseconnect/admin_import_export.py]: Double quotes please

File: nurseconnect/admin_views.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+from wagtailmodeladmin.views import IndexView
+
+# Custom - specific to NurseConnect. Adpated from the one
+# provided by molo.profiles
+from nurseconnect.admin_import_export import FrontendUsersResource
+
+
+class NurseConnectFrontendUsersAdminView(IndexView):
+    def post(self, request, *args, **kwargs):
+        drf__date_joined__gte = request.GET.get('drf__date_joined__gte')
```
Comment (chrislombaard) [inline nurseconnect/admin_views.py]: Double quotes please

File: nurseconnect/admin_views.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        drf__date_joined__lte = request.GET.get('drf__date_joined__lte')
```
Comment (chrislombaard) [inline nurseconnect/admin_views.py]: Double quotes please

File: nurseconnect/admin_views.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        is_active_exact = request.GET.get('is_active__exact')
```
Comment (chrislombaard) [inline nurseconnect/admin_views.py]: Double quotes

File: nurseconnect/admin_views.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+        filter_list = {
+            'date_joined__range': (drf__date_joined__gte,
```
Comment (chrislombaard) [inline nurseconnect/admin_views.py]: Double quotes for rest of function please.

File: nurseconnect/wagtail_hooks.py [hunk 15]
```diff
@@ -1,7 +1,28 @@
...
+
+def wagtailmodeladmin_register_without_menu(wagtailmodeladmin_class):
+    """
+    Based on wagtailmodeladmin_register which performs
+    three functions to ModelAdmin. Main problem with the
+    default behaviour is that
+    """
+    instance = wagtailmodeladmin_class()
+
+    @hooks.register('register_permissions')
```
Comment (chrislombaard) [inline nurseconnect/wagtail_hooks.py]: Double quotes in this file please

**Review by chrislombaard — CHANGES_REQUESTED** [review]
Just those small changes

File: nurseconnect/admin_import_export.py (hunk 8 above)
Comment (phala) [reply nurseconnect/admin_import_export.py:22]: I actually prefer the space to be there

File: nurseconnect/wagtail_hooks.py (hunk 15 above)
Comment (chrislombaard) [reply nurseconnect/wagtail_hooks.py]: Not these as well?

File: nurseconnect/wagtail_hooks.py [hunk 16]
```diff
@@ -1,7 +1,28 @@
...
+    def register_permissions():
+        return instance.get_permissions_for_registration()
+
+    @hooks.register('register_admin_urls')
```
Comment (chrislombaard) [inline nurseconnect/wagtail_hooks.py]: And these?

**Review by chrislombaard — CHANGES_REQUESTED** [review]

**Review by chrislombaard — APPROVED** [review]
+1

**Comment by chrislombaard** [comment]
+1

**Pull request merged**